Complete Apollo client integration, add FeatureFlag component

matthew-chirgwin opened the pull request:
- Fleshes out the config context to retrieve from the server, and stores the result in context
- Inserts bootstrap config and provides it via context
- Adds feature flagging components
- Tests for all areas


Known issues:

~400/404 errors will appear in the browser console as the UI attempts to connect to the `/api` endpoint. I have a fix for this, which will follow in another PR (pending #85)~ - in the interest of time, I have added a small change to make all GQL requests via HTTP - this stops errors from being thrown/enables UI development transpartent to the transport being used. I will follow up with a WS change in another PR

matthew-chirgwin committed:
feat: complete ConfigFeatureFlag context

- implements and tests ConfigFeatureFlag context
- adds new Apollo test helper test utility + docs
- fixes a few nits found while testing

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add Feature Flag component

- implements feature flag component + tests
- integrates (example) flag with Home
- adds context test helper

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: address review comments

- addresses review comments from #94

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
fix: renamed type

- fixes renamed type in test code
- moves runtime configuration pieces from serverConfig to config module

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,382 @@
+                  triggerRefetch,
+                  rawResponse,
+                }) => {
+                  expect(loading).toBe(true);
+                  expect(error).toBe(false);
+                  expect(isComplete).toBe(false);
+                  expect(rawResponse).toEqual({});
+                  expect(triggerRefetch).toEqual(expect.any(Function));
+                  return <p>{JSON.stringify(isComplete)}</p>;
```
Is there a world where these expects don't run so the test passes anyway? Is it worth passing the done parameter in so we know for sure that the test has completed because these asserts have happened?

matthew-chirgwin replied:
If the expect fails, the error bubbles up and fail the test - see https://github.com/strimzi/strimzi-ui/pull/94#discussion_r521996835

nictownsend replied:
Can you extract this into an "expectations" render function so it's not nested inside the component here

JQrdan commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,60 @@
+/** ConfigFeatureFlagConsumer - consumer component of the ConfigFeatureFlag context */
+const ConfigFeatureFlagConsumer = ConfigFeatureFlag.Consumer;
+/** Test provider. Only to be used in tests */
+const _ConfigFeatureFlag_TestProvider = ConfigFeatureFlag.Provider;
```
Code specific for tests in the real implementation? 😬

JQrdan commented on client/Contexts/ConfigFeatureFlag/README.md:
```diff
@@ -0,0 +1,31 @@
+# ConfigFeatureFlag
+
+This context will perform and store the result of a GraphQL query against the `/config` endpoint presented by the UI server, returning all configuration and feature flag state. This will allow the UI client discover and refresh itself at runtime. The context will provide hooks and consumers to direct access both the feature flag and config values.
```
```suggestion
This context will perform and store the result of a GraphQL query against the `/config` endpoint presented by the UI server, returning all configuration and feature flag state. This will allow the UI client to discover and refresh itself at runtime. The context will provide hooks and consumers to directly access both the feature flag and config values.
```

JQrdan commented on client/Contexts/ConfigFeatureFlag/README.md:
```diff
@@ -0,0 +1,31 @@
+
+- `ConfigFeatureFlagProvider` is a component which makes the request to the `/config` endpoint, and manages the result. It should only be placed in the `index.ts` in `Bootstrap` as it is intended to be a single use component.
+- `ConfigFeatureFlagConsumer` is a standard Context consumer, providing the `value` of the context at render time.
+- `useConfigFeatureFlag` is a hook, which provides the `value` of the context at render time.
```
it would be nice to provide an example of how to use the `useconfigFeatureFlag` hook here as well as the component

JQrdan commented on client/Panels/Home/Home.tsx:
```diff
@@ -4,12 +4,19 @@
 
 const Home: FunctionComponent = ({ children }) => {
+  const { client, isComplete } = useConfigFeatureFlag();
+
   return (
-    <div>
+    <div className='home'>
       <img src={image} alt='Strimzi logo' />
       Welcome to the Strimzi UI
+      <FeatureFlag flag={'client.Home.showVersion'}>
+        {isComplete && `Version: ${client.version}`}
+      </FeatureFlag>
```
I'm not so sure about the mixing of a hook and a component to achieve this. Is there a way where the hook itself can give you enough information so that a developer could just use that result to decide what to render?

JQrdan replied:
Sorry I've just noticed these are two separate ways of using this!

matthew-chirgwin replied:
Yeah - a user could use the hook, or the component, whatever makes sense. I can see cases where both would be helpful

matthew-chirgwin committed:
feat: add overrides to config, bootstrap config in html

- integrates the merging of overrides from server config to config
defined from the ui-config
- templates bootstrap config into html
- tests to cover

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,382 @@
+  });
+
+  describe('`useConfigFeatureFlag` hook', () => {
+    const HookWrapper: FunctionComponent = () => {
+      const { triggerRefetch, ...others } = useConfigFeatureFlag();
+      return (
+        <p>
+          {JSON.stringify({
+            ...others,
+            triggerRefetch: typeof triggerRefetch === 'function',
+          })}
+        </p>
+      );
+    };
```
Is there a reason you're doing it this way instead of using https://www.npmjs.com/package/@testing-library/react-hooks?

matthew-chirgwin replied:
Did not realise this was a thing! Will update to use it

JQrdan commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,60 @@
+
+const ConfigFeatureFlag = createContext(defaultConfigFeatureFlagValue);
+/** useConfigFeatureFlag - hook which returns the current value of the ConfigFeatureFlag context */
+const useConfigFeatureFlag: () => ConfigFeatureFlagType = () =>
```
Opinions on this being exposed through the hooks folder instead?

matthew-chirgwin replied:
Fair point - think this makes sense, and also means I can get rid of the 'private' provider as I will now need to export the context as a whole (this can use the provider from that). Will update

JQrdan commented on utils/test/context/context.util.tsx:
```diff
@@ -4,46 +4,75 @@
 
-export { renderWithContextProviders };
+/** withConfigFeatureFlagContext renders the given `ui` JSX with a ConfigFeatureFlag provider seeded with default values. Use this if your components indirectly use/require a value from the ConfigFeatureFlag context, else use `withCustomConfigFeatureFlagContext` so the value from ConfigFeatureFlag can be controlled  */
+const withConfigFeatureFlagContext: (
```
Given the naming of these functions, I'd assume that they would be used like:

```
render(withConfigFeatureFlagContext(...))
```

But the render function isn't needed. Could these be modified so they work in that way or renamed to `renderWithCustomConfigFeatureFlagContext` so that the names are consistent with the behaviour

matthew-chirgwin replied:
Agreed - will be updated in the next commit

matthew-chirgwin committed:
refactor: review comments, docs

- address review comments
- update docs

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+import { ConfigFeatureFlagType } from './ConfigFeatureFlag.types';
+
+describe('ConfigFeatureFlag', () => {
+  const mockClientResponse = {
```
Move to `assets`?

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+    },
+  ];
+
+  const noOpResponse = [
```
```suggestion
  const noOpQueryMock = [
```

matthew-chirgwin replied:
I am going to move this to assets, and also add a helper to avoid duplicating response structure

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+
+  describe('`ConfigFeatureFlagProvider` and `ConfigFeatureFlagConsumer` components', () => {
+    describe('initial state', () => {
+      it('returns the expected client configuration values', () => {
```
```suggestion
      it('returns the default client configuration values', () => {
```

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+      it('returns the expected client configuration values', () => {
+        const expectedValues = defaultClientConfig.client;
+        const { getByText } = render(
+          withApolloProviderReturning(
```
```suggestion
          setup(gqlMock, renderFn)
```
Where setup can create the `withApolloProvider` duplication

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+    });
+    describe('response handling logic', () => {
+      // define common expected states, namespaces so RTL getByText can identify them
+      const namespaceValue: (
```
Common `stringifyForRender` or similar?

matthew-chirgwin committed:
feat: add client bootstrap config to context

- updates Context to pull in config from HTML with tests

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
fix: paths on server

- fix path usage in server

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+          )
+        );
+        act(() => {
+          expect(getByText(JSON.stringify(expectedClient))).toBeInTheDocument();
```
https://testing-library.com/docs/dom-testing-library/api-async - `waitFor` of any use here?

matthew-chirgwin replied:
Aha! Yep, looks like a good fit, will update

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.spec.tsx:
```diff
@@ -0,0 +1,347 @@
+            errorResponse,
+            <ConfigFeatureFlagProvider>
+              <ConfigFeatureFlagConsumer>
+                {testRenderFunc}
```
If we control render function, is the `expect` inside render better than having to stringify and use `getByText`?

matthew-chirgwin replied:
Potentially - I went with this style given the setup/do/assert pattern where possible

nictownsend commented on client/Contexts/ConfigFeatureFlag/ConfigFeatureFlag.types.ts:
```diff
@@ -0,0 +1,31 @@
+import { exposedClientType, exposedFeatureFlagsType } from 'ui-config';
+
+/** the response (`data`) value from the graphql config query (client/Queries/Config/index.ts) */
+export type apolloQueryResponseType = {
```
General question - should we treat types like classes and have CamelCase as a naming strategy?

matthew-chirgwin replied:
I don't have a preference to which, but they should be consistent. Will update `apolloQueryResponseType` to be `ApolloQueryResponseType`

nictownsend commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,57 @@
+    featureFlags,
+    loading,
+    error: error ? true : false,
+    isComplete: !(loading || error) && !loading && !error,
```
```suggestion
    isSuccess: !(loading || error),
```

Might be going crazy but can't see why the second case is needed?

matthew-chirgwin replied:
I think you are right! Will remove the first case as feel the second is more readable. As for the name, would prefer (semantically) to keep the `isComplete` notion for readability of consumers of the context

nictownsend commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -0,0 +1,57 @@
+    loading,
+    error: error ? true : false,
+    isComplete: !(loading || error) && !loading && !error,
+    triggerRefetch: () => {
```
Why is this function nested and not just an alias for `refetch`?

matthew-chirgwin replied:
Waiting for logging line to be added. Otherwise, yes it could be

nictownsend commented on client/Contexts/ConfigFeatureFlag/FeatureFlag.view.tsx:
```diff
@@ -0,0 +1,28 @@
+  const featureFlagValue = get(featureFlags, flag, false);
+
+  return (
+    <div {...others} data-ff={flag}>
```
Is it more or less efficient to return null without the `<div>` when the flag is disabled?

matthew-chirgwin replied:
From a rendered elements POV yes - I was speculating on the helpfulness is debug/test to confirm the container had no content, ie the flag was off

matthew-chirgwin commented:
This PR now rebased on master - will rebase again once #85 goes in as will need some of it's WS work to integrate Apollo fully

nictownsend commented on client/Panels/Home/Home.steps.tsx:
```diff
@@ -15,12 +16,25 @@ Given('a Home component', () => {
 
 When('it is rendered', () => {
-  renderResult = render(component);
+  renderResult = renderWithCustomConfigFeatureFlagContext(
```
Can we have a separate test with flags disabled?

nictownsend replied:
https://github.com/kentcdodds/react-test-context-provider?

nictownsend replied:
Also raises an interesting question - should components with context be tested without a context (to check default behaviour)?

And I'm guessing it means that we have to have a mixture of `prop` and `context` test cases?

matthew-chirgwin replied:
Will add a test with flags off now.

Interesting question re testing without context. Contexts will always return a default value - I would tend to have a test which perhaps (via `renderWithConfigFeatureFlagContext`) checks behaviour of this component using the context with default values.

nictownsend commented on utils/test/context/context.util.tsx:
```diff
@@ -36,4 +42,37 @@ const renderWithContextProviders: <T>(
+  ]);
+
+/** renderWithCustomConfigFeatureFlagContext renders the given `ui` JSX with a ConfigFeatureFlag provider provided via `configFeatureFlagValue`. */
+const renderWithCustomConfigFeatureFlagContext: (
```
Can these be composed? I.e can I perform `renderWithConfigFeature(renderWithCustomConfig)` - or a builder esque pattern that allows you to do `withX().withY().withZ().render()`?

matthew-chirgwin replied:
Agree it would be nice to have something more fluent, but I was building upon what was already there. Follow on issue?

nictownsend commented on client/Panels/Home/Home.tsx:
```diff
@@ -3,14 +3,22 @@
+import { useConfigFeatureFlag } from 'Hooks';
 
 const Home: FunctionComponent = ({ children }) => {
+  const { client, featureFlags, isComplete } = useConfigFeatureFlag();
```
Is the design that any component wanting to use flags will need loading indicators?

matthew-chirgwin replied:
As they are retieved from the server, yes, they may need to. This, along with introspection and authz (if needed) would be covered (I suggest) under a single loading state 'on mount'/bootstrap , covered by UX issue #87

nictownsend commented on client/Queries/Config/index.ts:
```diff
@@ -0,0 +1,25 @@
+
+export const GET_CONFIG = gql`
+  query {
+    client {
```
Semantically, `about` doesn't feel like "config" - is this more designed to get "runtime" info?

matthew-chirgwin replied:
I wanted to provide an example of configuration (settings or values) being used in this PR. All of these settings or values would be returned at runtime - I don't mind the name changing if you can think of something better! :)

pmuir replied:
I think the config name is fine and fairly standard.

nictownsend commented on client/Queries/Config/index.ts:
```diff
@@ -0,0 +1,25 @@
+    }
+    featureFlags {
+      client {
+        Home {
```
Should this be a result of introspection? Or alternatively, in gql something like

`client: [featureFlag]`
`type featureFlag = {name:value}`

Perhaps with nesting (so we can define a shape)?

matthew-chirgwin replied:
Introspection is of the Strimzi-admin capability, and if something is enabled (or not) - this is UI capability. So similar, but different.

Re the typing/structure - as the schema is defined dynamically from the config, this is what we get OOTB. I would suggest this is ok as is, but could be changed in future if needed.

pmuir replied:
Agreed, feature flags are orthogonal to introspection.

nictownsend commented on config/featureflags.ts:
```diff
@@ -2,10 +2,19 @@
-export const featureFlags: Config<boolean> = {};
+export const featureFlags: Config<boolean> = {
+  client: {
+    Home: {
```
Same concern over having to define these (they're more like `ConfigValue<boolean>`)

nictownsend replied:
It feels like the UI should not have to know about the flag structure to query for it - and same with server, it parses some feature flag config and then has a key:value structure rather than having to define each feature flag.

pmuir replied:
I think it's better have an explicit schema for the feature flags (which I think this is providing). A freeform map is quite hard to manage in practice, as you don't know the total set of flags you have in the app.

nictownsend commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "description": "Strimzi UI provides a way for managing Strimzi and Kafka clusters (+ other components) deployed by it using a graphical user interface.",
   "scripts": {
-    "clean": "rm -rf dist/ && rm -rf js/ && rm -rf .out/ && rm -rf generated/ && rm -rf coverage/",
+    "clean": "rm -rf dist/ && rm -rf js/ && rm -rf .out/ && rm -rf generated/ && rm -rf coverage/ && npx jest --clear-cache",
```
Assumes `npx` is installed. You can instead have a `npm run clean-cache` that calls `jest --clear-cache`

matthew-chirgwin replied:
Not an issue as it is included with the minimum version of npm we reference as a requriement

nictownsend commented on server/client/router.ts:
```diff
@@ -31,6 +32,13 @@ export const ClientModule: UIServerModule = {
+    hasIndexFile &&
+      routerForModule.get(
+        '/index.html',
+        renderIndexHtmlWithBootstrapConfig(indexFile)
```
`renderTemplate` maybe is less verbose?

nictownsend commented on server/config/controller.ts:
```diff
@@ -3,48 +3,65 @@
+export const returnSchemaAndResolversGivenConfig: (
+  config: serverConfigType
+) => configSchemaResolverType = (config) => {
+  const { featureFlags: featureFlagOverrides } = config;
```
```suggestion
  const { featureFlags: featureFlagOverrides, client: {configOverrides} } = config;
```

nictownsend commented on server/config/controller.ts:
```diff
@@ -3,48 +3,65 @@
-            ...allResolvers,
+  const configToHost = {
+    featureFlags: merge(
+      merge({}, featureFlags.publicValues),
```
Why the extra `merge`? `merge` is variadic so you can do `merge({}, x, y, z)`

matthew-chirgwin replied:
Aha! Much nicer! Will update (along with others)

nictownsend commented on server/config/router.ts:
```diff
@@ -7,17 +7,19 @@ import { UIServerModule } from 'types';
 
-    const server = new ApolloServer({ typeDefs: schema, resolvers });
+    const server = new ApolloServer({
+      ...returnSchemaAndResolversGivenConfig(config),
```
```suggestion
      ...apolloConfig(config),
```

matthew-chirgwin committed:
feat: add URL feature flag overrides

- adds feature flag overrides from URL + doc + tests
- adds new utils to enable this (input sanitisation, window interaction)

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: address review comments

- addresses review comments from @nictownsend

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

ajborley commented on client/Contexts/ConfigFeatureFlag/Context.tsx:
```diff
@@ -2,41 +2,95 @@
+    data || defaultClientConfig;
+
+  // check to see/merge any feature flag state with any defined in browser
+  const sanatisedParams = sanatiseUrlParams(getLocation().search);
```
Typo: `sanitised`
```suggestion
  const sanitisedParams = sanitiseUrlParams(getLocation().search);
```

matthew-chirgwin replied:
Thanks Andy - will fix (and rename all files etc) in a new commit

matthew-chirgwin committed:
fix: spelling error, package rebase error

- reverts mini css extract plugin to match master (changed in error in
rebase)
- fixes spelling of sanitise

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: move GQL traffic to HTTP

- move all GQL requests to be HTTP based, with Websocket support to
follow

Contributes to: #13

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented:
Thank you for reviews/comments @JQrdan @dlabaj  @nictownsend @ajborley - I have just pushed a commit to change the transport to HTTP for now, while I work on the WS support. Can I please ask for another review, with a view to merge this asap as long as there are no objections?

JQrdan reviewed (approved):
HTTP changes also look good to me

pmuir reviewed (approved):
LGTM - note I only did a review of the semantics of the feature flags and apollo client integration, I didn't attempt to a code review.
